### ISSUE TITLE
Consider designing a runtime room data disk cache

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#731]
As of now the rooms keep their saved state in memory. As player visits more rooms their data in memory increases. I do not know if that was ever a real problem, but it's possible to assume that at some point it may be.

Internally this is managed by array of RoomStatus objects.
IIRC what is kept:
* room script variables;
* room object properties (RoomObject class);
* custom property values (added in 3.4.0);
* hotspot/areas enabled states;
* some minor additional data.

Partially this situation is eased by room backgrounds discarding any changes when room is unloaded, so engine does not have to keep modified copies of backgrounds in memory. I have a guess this was done exactly to prevent memory issues.

What I'd like to suggest, at least for consideration, is the runtime disk cache system. Instead of keeping all room states in memory at all times engine could only keep limited size of data (configurable amount) and dump the rest to disk in user files directory (e.g. near savegames). As rooms are visited again or data has to be accessed for any reason engine would restore this data from the disk.

**Comment by rofl0r** [comment]
my suggestion is to benchmark the actual memory consumption on a number of games so you have some hard data to base your decisions on.

**Comment by ivan-mogilko** [comment]
Yes, the problem may be theoretical atm, but it was bothering me for some time so I decided to open this issue for the record.

Something that I forgot to mention (added to first post now), there is also a possibility that room script allocates large amount of managed objects. If these objects are only referenced from the room script when it's unloaded, then they will stay in memory for no good (even if room is not visited ever again).

**Comment by ericoporto** [comment]
This is a pretty interesting proposal, but since the Precache API is something that begins to exist (for sprites), it appears this would benefit from such an API, so that when you load a room you can manually tell ags about the adjacent ones, in a non-blocking way, so it doesn't slow it down. Alternatively, the room itself could have some property where you can add to a list (possibly through the Editor) specifying what are the adjacent rooms. Rooms that aren't the one loading would have to be loaded in a non-blocking way just to make it not slow on room transition.
